What's New 93 (#1364)
MicrosoftDocs/edge-developer: microsoft-edge/devtools-guide-chromium/whats-new/2021/07/devtools.md

---
description: Apply color themes from Visual Studio Code, debug DOM node memory leaks with the new Detached Elements tool, Microsoft Edge Developer Tools for Visual Studio Code now integrates with the Visual Studio Code Debugger workflow, and more.
title: What's new in DevTools (Microsoft Edge 93)
author: MSEdgeTeam
ms.author: msedgedevrel
ms.date: 07/30/2021
ms.topic: article
ms.prod: microsoft-edge
keywords: microsoft edge, web development, f12 tools, devtools
---
# What's New In DevTools (Microsoft Edge 93)

[!INCLUDE [contact DevTools team note](../../includes/edge-whats-new-note.md)]

<!-- 1 -->
## Apply themes from Visual Studio Code to DevTools

<!-- Title: Apply themes from Visual Studio Code to DevTools -->
<!-- Subtitle: You can now use some of the most popular color themes from Visual Studio Code, such as Monokai and Solarized Dark, in Microsoft Edge DevTools. -->

In addition to the existing light and dark themes, Microsoft Edge DevTools now supports some of the most popular color themes from Visual Studio Code.  To select a color theme, navigate to **Settings** and then select a theme from the **Theme** dropdown list.

> :::image type="complex" source="../../media/2021/07/all-devtools-themes.msft.png" alt-text="Color themes for DevTools" lightbox="../../media/2021/07/all-devtools-themes.msft.png":::

Author: tried quick fix: Paint 3D > Canvas > Transparent canvas.

## Debug DOM node memory leaks with the new Detached Elements tool

:::image type="icon" source="../../media/2020/06/experimental-tag-14px.msft.png":::  

<!-- Title: Introducing the Detached Elements tool -->
<!-- Subtitle: Use the Detached Elements tool to find and fix DOM node memory leaks. -->

A DOM node is considered "detached" when it is no longer attached to any element of the DOM but is still being retained in memory by Microsoft Edge. The browser cannot garbage-collect the element, because some JavaScript is still referencing the element even though it's no longer on the page or is not longer a part of the DOM.

The new **Detached Elements** tool finds all of the detached elements on your page and displays them. You can expand and collapse a detached element to see the parent and child nodes that are also being retained. You can trigger the browser's garbage collection by selecting **Collect garbage** and validate that you have a memory leak when a detached element cannot be garbage-collected. And you can jump into JavaScript that's referencing the detached element by taking a heap snapshot with the **Analyze** button.

<!-- todo: comment-out or delete steps? -->

> 1. In Microsoft Edge Canary, navigate to [Microsoft ToDo](https://to-do.office.com/tasks/inbox).

Reviewer: It wasn't clear to me at first that the numbered steps were for a tutorial. Maybe it would be helpful to add a sentence before the list that says something like "A good way to test the Detached Elements tool is by debugging a to-do list web application such as Microsoft To Do." That way, it's clear that the steps are a tutorial and not required setup instructions.

1. Select the **+** next to **Add a task**, and then select **Add**.
1. Select the task you created, select **Add due date**, select **Pick a date**, and then select a date from the calendar.  Then select **Save**.
1. In DevTools, in the **Detached Elements** tool, select **Get Detached Elements**.
1. If the **Analyze** button is grayed-out, add another task and date, then select **Get Detached Elements** again.<!--added extra step-->
1. Select the **Analyze** button in the toolbar and wait for the **Id** column to populate.
1. In the **Id** column, select an ID.  The **Memory** tool opens in the bottom panel, and displays the retainers list.

:::image type="complex" source="../../media/2021/07/detached-elements-tool.msft.png" alt-text="The Detached Elements tool" lightbox="../../media/2021/07/detached-elements-tool.msft.png":::
   The **Detached Elements** tool
:::image-end:::

For more information, navigate to the **Detached Elements** section in [Experimental features](../../../experimental-features/index.md).
<!-- todo: link to subheading in page, when available -->

<!-- no crbug -->

<!-- 3 -->

> ## Microsoft Edge Developer Tools for Visual Studio Code now integrates with the Visual Studio Code Debugger workflow

Author: I have seen a style guide rule somewhere that says a heading doesn't have to spell out the full formal name of a product, as long as first-mention in body gives the proper name.  I'll add some commented-out variant candidate headings.

<!-- Title: While debugging, launch the DevTools extension by selecting the Inspect button -->
<!-- Subtitle: Microsoft Edge DevTools for Visual Studio Code now integrates seamlessly with the JavaScript debugging workflow in the editor. -->

If you use the JavaScript debugging workflow in Visual Studio Code, you can now launch the DevTools extension by selecting the **Inspect** button.

> :::image type="complex" source="../../media/2021/07/inspect-button.msft.png" alt-text="The Inspect button in Visual Studio Code to launch the DevTools extension" lightbox="../../media/2021/07/inspect-button.msft.png":::

Author: @Reezaali - I checked "images" in Doc Contributor Guide, don't see a way to center a ::: image.  Are some ::: images centered somewhere in DMC?

> :::image type="complex" source="../../media/2021/07/inspect-button.msft.png" alt-text="The Inspect button in Visual Studio Code to launch the DevTools extension" lightbox="../../media/2021/07/inspect-button.msft.png":::

Author: By Design.  I don't think there's a way to center a :::-tagged image.
- I searched the repo for "center", not found in the context of centering an image.
- I read the Docs Contributor Guide re: :::-tagged images; the word 'center' doesn't appear.
- I assume all images in this repo are left-justified, never centered.  I'll keep an eye out for centered images.

   The **Inspect** button in Visual Studio Code to launch the DevTools extension
:::image-end:::

> This feature integrates DOM and CSS debugging with the JavaScript debugging workflow in Visual Studio Code.  If you don't have the DevTools extension installed, when you select the **Inspect** button, Visual Studio Code prompts you to install the extension seamlessly.

Reviewer 2: ```suggestion
This feature integrates DOM and CSS debugging with JavaScript debugging in Visual Studio Code.  If you don't have the DevTools extension installed, when you select the **Inspect** button, Visual Studio Code prompts you to install the extension.
```

> Other new features are:

Reviewer 2: I feel like the order of this list should be reversed. So it should be:

- The tools refresh automatically when you switch between different debugging targets.
- Several bug fixes.
- More detailed documentation of the extension.

*  More detailed documentation of the extension.
*  Several bug fixes.
*  The tools refresh automatically when you switch between different debugging targets.

For more details about improvements and fixes, check the changelog file in the [vscode-edge-devtools repo](https://github.com/microsoft/vscode-edge-devtools/blob/main/CHANGELOG.md).

:::image type="complex" source="../../media/2021/07/extension-integrated-debugger.msft.png" alt-text="DevTools extension integrated with Visual Studio Code Debugger workflow" lightbox="../../media/2021/07/extension-integrated-debugger.msft.png":::
   DevTools extension integrated with Visual Studio Code Debugger workflow
:::image-end:::

> For more information, navigate to [Launching Edge DevTools from the JS Debugger workflow][GithubVscodeEdgeDevtoolsDebuggerIntegration].

Reviewer 2: Can we combine this and the next 3 lines into one paragraph?

You can file issues and contribute to the extension on the [vscode-edge-devtools GitHub repo][GithubMicrosoftVscodeEdgeDevtools].

<!-- 4 -->
## New Fluent UI icons for DevTools

<!-- Title: New look for buttons and menus in Microsoft Edge DevTools -->
<!-- Subtitle: DevTools has adopted Fluent UI, giving it a more modern look that better aligns with the rest of the Microsoft Edge browser. -->

Microsoft Edge DevTools has adopted [Fluent UI](https://developer.microsoft.com/en-us/fluentui#/), giving buttons and menus a more modern look that better aligns with the rest of the Microsoft Edge browser.

:::image type="complex" source="../../media/2021/07/fluent-ui.msft.png" alt-text="DevTools implemented with Fluent UI design" lightbox="../../media/2021/07/fluent-ui.msft.png":::
   DevTools implemented with Fluent UI design
:::image-end:::

<!-- no crbug -->

<!-- 5 -->
## Change the DevTools display language directly from Settings

<!-- Title: DevTools Settings now includes display language -->
<!-- Subtitle: You can now skip the browser settings and change the DevTools display language directly within DevTools Settings. -->

> Previously, to change the display language in DevTools, you had to change the browser language.  Now you can easily switch the locale in DevTools **Settings**, without having to change your browser settings.  To do this, open **Settings**, and then in **Preferences**, select a language from the **Language** drop-down list.

Reviewer 2: ```suggestion
Previously, to change the display language in DevTools, you had to change the browser language.  Now you can easily switch the display language in DevTools **Settings**, without having to change your browser settings.  To do this, open **Settings**, and then in **Preferences**, select a language from the **Language** drop-down list.
```

:::image type="complex" source="../../media/2021/07/settings-browser-ui-language.msft.png" alt-text="Changing the DevTools display language directly from DevTools **Settings**" lightbox="../../media/2021/07/settings-browser-ui-language.msft.png":::
   Changing the DevTools display language directly from DevTools **Settings**
:::image-end:::

> For options for changing settings, navigate to [Customize Microsoft Edge DevTools](../../../customize/index.md).  To review the history of this feature in the Chromium open-source project, navigate to Issue [2882756][CR2882756].

Reviewer 2: ```suggestion
By default, the DevTools will match your browser language. For more customization options in DevTools, navigate to [Customize Microsoft Edge DevTools](../../../customize/index.md).  To review the history of this feature in the Chromium open-source project, navigate to Issue [2882756][CR2882756].
```

Reviewer 2: @erdraud, please put up a PR that updates https://docs.microsoft.com/en-us/microsoft-edge/devtools-guide-chromium/customize/localization as well so that we can link there from this announcement instead of the more general Customize overview

Author: I'm not committing the above Suggestion; reworded locally.

User-Agent Client Hints makes browser information more accessible than a semicolon-delimited User-Agent string, and improves site compatibility.  Initially, User-Agent Client Hints were time-consuming to test and debug.  There was less control over the client hints, and client hints had to be filled-in properly for the form to work.

In this release, we redesigned the debugging experience so you can easily modify User-Agent Client Hints through a UI with multiple separate fields and controls.  Also, you can now test your custom User-Agent Client Hints and a User-Agent string simultaneously.  You can now define User-Agent Client Hints for a custom device by using DevTools **Settings** or by using the **Network conditions** tool.

To define User-Agent Client Hints for a custom device by using DevTools **Settings**:
1. Navigate to **DevTools** > **Settings** \(![Settings icon](../../../media/settings-icon-dark.msft.png)\) > **Devices**.
1. In the **Emulated Devices** settings page, select the **Add custom device** button.
1. Expand the **User agent client hints** section.

> 1. Fill-in some of the fields.

Reviewer: ```suggestion
1. Fill in some of the fields.
```

:::image type="complex" source="../../media/2021/07/ua-client-hints-in-settings.msft.png" alt-text="Defining User-Agent Client Hints for a custom device in Settings" lightbox="../../media/2021/07/ua-client-hints-in-settings.msft.png":::
   Defining User-Agent Client Hints for a custom device in **Settings**
:::image-end:::

> To define User-Agent Client Hints for a custom device by using the **Network conditions** tool:

Reviewer 3: ```suggestion
To override User-Agent Client Hints for the current page by using the **Network conditions** tool:
```

> 1. In **DevTools**, select the **Network conditions** tool.  You might need to select **More Tools** (+) and then select **Network conditions**.

Reviewer 2: @erdraud, I think this is another one where we could move these steps into a new doc or update ones like this: https://docs.microsoft.com/en-us/microsoft-edge/devtools-guide-chromium/device-mode/override-user-agent and that would let us keep the announcement a bit shorter. We could just have the 2 images in a 2 column format showing the new UI in Settings and Network conditions.

If you agree with this plan, please put up a PR that either adds a new doc or updates an existing one for this feature

Author: I removed the steps.  Probably no other action is needed, because an equivalent PR has been merged already; the updated articles already have the equivalent writeup of the steps:
o	https://docs.microsoft.com/en-us/microsoft-edge/devtools-guide-chromium/device-mode/#set-user-agent-client-hints 
o	https://docs.microsoft.com/en-us/microsoft-edge/devtools-guide-chromium/network/reference#set-user-agent-client-hints
Step 1 means right-click _in the rendered webpage_ and then select Inspect.
o	Related: https://docs.microsoft.com/en-us/microsoft-edge/devtools-guide-chromium/device-mode/override-user-agent  - Override the user agent string from Microsoft Edge DevTools

1. In the **User agent** row, clear the **Use browser default** checkbox.
1. Expand the **User agent client hints** section.

> 1. Keep **Custom** selected in the drop-down list, and fill-in some of the fields.  Or, in the drop-down list, select an existing device, to automatically fill-in the fields based on an existing device.

Reviewer: ```suggestion
1. Keep **Custom** selected in the drop-down list, and fill in some of the fields.  Or, in the drop-down list, select an existing device, to automatically fill in the fields based on an existing device.
```

Author: Confirmed that "fill in the" is usual.  Will remove hyphen. thx

:::image type="complex" source="../../media/2021/07/ua-client-hints-in-network-conditions.msft.png" alt-text="Defining User-Agent Client Hints for a custom device in the Network conditions tool" lightbox="../../media/2021/07/ua-client-hints-in-network-conditions.msft.png":::
   Defining User-Agent Client Hints for a custom device in the **Network conditions** tool
:::image-end:::

> For more information about this feature, navigate to [Detecting Microsoft Edge from your website][DevToolsWebPlatformUserAgentGuidance].<!--no images?-->

Reviewer 2: Can we combine this sentence and the next one into one paragraph?

<!-- optional todo: compare the What's New screenshots vs the screenshots in these two incoming articles: -->
<!-- todo: link when this PR is merged:
For more information about this feature, navigate to [Emulate mobile devices in Microsoft Edge DevTools](../../../device-mode/index.md#set-user-agent-client-hints) and [Network Analysis reference](../../../network/reference.md#set-user-agent-client-hints). -->

To review the history of this feature in the Chromium open-source project, navigate to Issue [1174299][CR1174299].

<!-- 8 -->
## Screen readers now announce errors, warnings, and issues in toolbar and Console

<!-- Title: Better support for errors, warnings, and issues with assistive technology -->
<!-- Subtitle: Screen readers now correctly announce the number and the type of notification for errors, warnings, and issues in the DevTools toolbar. -->

> Previously, users of screen readers would only hear the number of errors, warnings, or issues announced in the DevTools toolbar.  The additional information of what type of notification was being announced wasn't included, such as "Error, Warning, or Issue". For example, if the DevTools reported 3 errors, screen readers would just announce "3".

Reviewer 2: ```suggestion
Previously, users of screen readers would only hear the number of errors, warnings, or issues announced in the DevTools toolbar.  The additional information of what type of notification was being announced wasn't included, such as "Error", "Warning", or "Issue". For example, if the DevTools reported 3 errors, screen readers would just announce "3".
```

Now in Microsoft Edge version 93, screen readers correctly announce the number of errors, warnings, or issues that the DevTools have detected and the type of notification. For example, if the DevTools reported 3 errors, screen readers now announce "3 errors".  This fix has been applied to both the notifications in the DevTools toolbar and in the Console.

:::image type="complex" source="../../media/2021/07/screen-reader-errors-warnings-issues.msft.png" alt-text="The errors, warnings, and issues UI in the toolbar and Console" lightbox="../../media/2021/07/screen-reader-errors-warnings-issues.msft.png":::
   The errors, warnings, and issues UI in the toolbar and **Console**
:::image-end:::

<!-- It'd be good to have a video of this a11y fix where the text that the screen reader announces is displayed -->

> For information about debugging Console errors, navigate to [Debug errors reported in Console](../../../console/console-debug-javascript.md).  For information about issues found by DevTools and improvements you can make to a webpage, navigate to [Find and fix problems using the Issues tool](../../../issues/index.md).

Reviewer 2: Can we combine this line with the next sentence so that we only have one paragraph?

<!-- Title: Generate PowerShell commands for network requests in the Network tool -->
<!-- Subtitle: The Copy as PowerShell context menu option now correctly sets the user-agent string and cookies when generating PowerShell network requests. -->

Previously, in the **Network** tool, the **Copy** > **Copy as PowerShell** context menu option didn't include cookies when generating a PowerShell command for a given network request in the Network Activity Log. This meant that the generated PowerShell command couldn't successfully make the same network request if cookies were required.

Now in Microsoft Edge version 93, the **Copy as PowerShell** context menu option correctly sets the User-Agent string and cookies that were observed by DevTools.  The generated PowerShell command can now successfully make the same network request that was observed by DevTools, even to servers that depend on cookies.

:::image type="complex" source="../../media/2021/07/copy-as-powershell.msft.png" alt-text="The Copy as PowerShell command" lightbox="../../media/2021/07/copy-as-powershell.msft.png":::
   The **Copy as PowerShell** command
:::image-end:::

> For more information about the Network Activity Log, navigate to [Network Analysis reference](../../../network/reference.md).

Reviewer 2: Same comment as above. Can we merge this sentence and the next one into one paragraph?
